rfcs/0014-opendal.md

```diff
@@ -0,0 +1,196 @@
+
+1. `put_opts()` with `PutMode::Create` for writing new versioned manifests with CAS semantics
+2. `get()` for reading specific manifest versions
+3. `list()` for listing manifest versions in a range
+
+This component provides the versioned storage protocol for `ManifestStore` and is particularly critical because it **uses `PutMode::Create` for optimistic concurrency control**. This create-if-not-exists semantics prevents concurrent manifest corruption, making it essential for fencing in distributed scenarios.
+
+The migration requires that OpenDAL **supports CAS (create-if-not-exists) semantics equivalent to `PutMode::Create`**—this is a P0 validation requirement. Additionally, the list operation must support range queries over versioned objects.
+
+<!-- TOC --><a name="retryingobjectstore"></a>
+### RetryingObjectStore
+
+In the short term, we can simply wrap the current implementation of `RetryingObjectStore` before passing `Arc<dyn object_store::ObjectStore>` to OpenDAL's operator.
+
+However, we should also ensure it's technically possible to replace the current implementation with OpenDAL's layers.
+
+The current `RetryingObjectStore` (`retrying_object_store.rs`) wraps any `ObjectStore` with exponential backoff retry logic. The implementation avoids retrying certain error types that indicate non-transient failures: `AlreadyExists`, `Precondition`, and `NotImplemented` errors are never retried to preserve correctness semantics.
+
+One notable implementation detail is the special handling of `list()` operations. Because the list API returns a stream and cannot be easily retried mid-iteration, the current implementation collects the entire stream into a Vec before attempting the operation. This ensures atomicity—either the entire list succeeds or the entire operation is retried.
+
+OpenDAL provides a [`RetryLayer`](https://opendal.apache.org/docs/rust/opendal/layers/struct.RetryLayer.html) with configurable retry policies that can potentially replace our custom implementation. It retries on errors which `Error::is_temporary` returns true.
+
+However, we may want to customize the retry behavior to match our current implementation. For example, when encountering an `AlreadyExists` error, we may want to read the object metadata (via a `stat()` call) to determine whether to return success (if we created the object and it already exists, making the operation idempotent) or return a conflict error (if someone else created it).
+
+Since OpenDAL's built-in `RetryLayer` did not expose an interface to allow us customize the retry behavior, we could consider implementing a customized `RetryLayer` in our repository. This approach would allow us to preserve our specialized retry logic (such as the idempotent `AlreadyExists` handling) while still benefiting from OpenDAL's ecosystem and avoiding the need to implement the entire `ObjectStore` trait as we do currently.
+
```

Comment:
During migration, I think we will now have three retries on top of eachother:

1. `object_store` has its own internal retries
2. `RetryingObjectStore` wraps an `object_store` and does still more retries
3. `opendal` does its own retries

I'm curious about (3): are retries in OpenDAL disabled by default? If not, do we want to disable all OpenDAL retries since we're getting (1) and (2) already?